salt/netapi/rest_tornado/saltnado.py

```diff
@@ -925,64 +947,83 @@ def _disbatch_local(self, chunk):
+        all_return_future = self.all_returns(pub_data['jid'],
+                                             is_finished,
+                                             minions_remaining=list(minions_remaining),
+                                             )
+        yield job_not_running_future
+        raise tornado.gen.Return((yield all_return_future))
 
     @tornado.gen.coroutine
     def all_returns(self,
                     jid,
-                    finish_futures=None,
+                    is_finished,
                     minions_remaining=None,
                     ):
         '''
         Return a future which will complete once all returns are completed
-        (according to minions_remaining), or one of the passed in "finish_futures" completes
+        (according to minions_remaining), or one of the passed in "is_finished" completes
         '''
-        if finish_futures is None:
-            finish_futures = []
         if minions_remaining is None:
             minions_remaining = []
 
-        ret_tag = tagify([jid, 'ret'], 'job')
         chunk_ret = {}
+
+        minion_events = {}
+        for minion in minions_remaining:
+            tag = tagify([jid, 'ret', minion], 'job')
+            minion_event = self.application.event_listener.get_event(self,
+                                                                     tag=tag,
+                                                                     matcher=EventListener.exact_matcher,
+                                                                     timeout=self.application.opts['timeout'])
+            minion_events[minion_event] = minion
+
         while True:
-            ret_event = self.application.event_listener.get_event(self,
-                                                      tag=ret_tag,
-                                                      )
-            f = yield Any([ret_event] + finish_futures)
-            if f in finish_futures:
-                raise tornado.gen.Return(chunk_ret)
-            event = f.result()
-            chunk_ret[event['data']['id']] = event['data']['return']
-            # its possible to get a return that wasn't in the minion_remaining list
+            f = yield Any(minion_events.keys() + [is_finished])
+            try:
+                if f is is_finished:
+                    for event in minion_events:
+                        if not event.done():
+                            event.set_result(None)
+                    raise tornado.gen.Return(chunk_ret)
+                f_result = f.result()
+                chunk_ret[f_result['data']['id']] = f_result['data']['return']
+            except TimeoutException:
+                pass
+
+            # clear finished event future
             try:
-                minions_remaining.remove(event['data']['id'])
-            except ValueError:
+                minions_remaining.remove(minion_events[f])
+                del minion_events[f]
+            except ValueError as e:
```

Comment:
Since you don't make use of `e`, perhaps just have:
```
except ValueError:
```

Reply:
this has been fixed